edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/Handlers/SubscriptionChangeHandler.cs

```diff
@@ -0,0 +1,119 @@
+    {
+        const string SubscriptionChangePattern = @"^\$edgehub/(?<id1>[^/\+\#]+)(/(?<id2>[^/\+\#]+))?/subscriptions$";
+
+        readonly IConnectionRegistry connectionRegistry;
+        readonly IComponentDiscovery components;
+        readonly IIdentityProvider identityProvider;
+
+        public SubscriptionChangeHandler(IConnectionRegistry connectionRegistry, IComponentDiscovery components, IIdentityProvider identityProvider)
+        {
+            this.connectionRegistry = connectionRegistry;
+            this.components = components;
+            this.identityProvider = identityProvider;
+        }
+
+        public async Task<bool> HandleSubscriptionChangeAsync(MqttPublishInfo publishInfo)
+        {
+            var match = Regex.Match(publishInfo.Topic, SubscriptionChangePattern);
+
+            if (!match.Success)
+            {
+                return false;
+            }
+
+            // After this point we know that this was a subscription change and will return 'true'
+            // even in a case of error
+            var subscriptionList = default(List<string>);
+            try
+            {
+                var payloadAsString = Encoding.UTF8.GetString(publishInfo.Payload);
+                subscriptionList = JsonConvert.DeserializeObject<List<string>>(payloadAsString);
+
+                Events.HandlingSubscriptionChange(payloadAsString);
+            }
+            catch (Exception e)
+            {
+                Events.BadPayloadFormat(e);
+                return true;
+            }
+
+            var id1 = match.Groups["id1"];
+            var id2 = match.Groups["id2"];
+
+            var identity = HandlerUtils.GetIdentityFromMatch(id1, id2, this.identityProvider);
+
+            var proxy = default(IDeviceListener);
+            var proxyMaybe = await this.connectionRegistry.GetUpstreamProxyAsync(identity);
+
+            if (!proxyMaybe.HasValue)
+            {
+                return true;
+            }
+            else
+            {
+                proxy = proxyMaybe.Expect(() => new Exception($"No upstream proxy found for {identity.Id}"));
+            }
+
+            foreach (var watcher in this.components.SubscriptionWatchers)
+            {
+                foreach (var subscriptionPattern in watcher.WatchedSubscriptions)
+                {
+                    var subscribes = false;
+
+                    foreach (var subscription in subscriptionList)
+                    {
+                        var subscriptionMatch = Regex.Match(subscription, subscriptionPattern.Pattern);
+                        if (subscriptionMatch.IsMatchWithIds(id1, id2))
+                        {
+                            subscribes = true;
+                            break;
+                        }
+                    }
+
+                    try
+                    {
+                        await proxy.AddOrRemoveSubscription(subscribes, subscriptionPattern.Subscrition);
```

Comment:
So, here it seems like we will handle all subscriptions on a subscription change? Can we instead only handle changed subscriptions?

Reply:
There was an option to maintain a state, so at this level we would know that when the broker sends an updated subscription list, that is a change for a given subscription or not - but this state is also maintained by SubscriptionProcessor/ConnectionManager. I didn't want to duplicate this state (afraid of getting out of sync and causing problems). Instead I wanted ConnectionManager to be the source of truth. There is a related change that ConnectionManager reports back that it was an actual change or not, so the subscription processor will not do unnecessary updates at IoT Hub. With that change, the "unnecessary" calls don't seem that big performance hit, also subscriptions don't happen often.